### PR TITLE
Save diamond conversion as bytes bzip2

### LINKED ISSUE #422: saveAsJSON in DiamondTabularFormatReader needs to know whether output is bzip2'd
If `convert-diamond-to-json.py` is told to produce bzip2 output, it needs to tell `saveAsJSON` to write out bytes instead of strings, otherwise the following error is raised:

```
Traceback (most recent call last):
  File "/scratch/tcj25/root/.virtualenvs/35/bin/convert-diamond-to-json.py", line 45, in <module>
    reader.saveAsJSON(fp)
  File "/scratch/tcj25/root/.virtualenvs/35/lib/python3.5/site-packages/dark/diamond/conversion.py", line 114, in saveAsJSON
    print(dumps(self.params, separators=(',', ':')), file=fp)
  File "/scratch/tcj25/root/.virtualenvs/35/lib/python3.5/site-packages/bz2file.py", line 388, in write
    compressed = self._compressor.compress(data)
TypeError: a bytes-like object is required, not 'str'
```

### PR DESCRIPTION
**Pull request by terrycojones**
This is a bit messy due to differences between Python 2 and 3.

Fixes #422.

### CODE REVIEW

**Comment by terrycojones** [comment]
BTW, I changed `50-60` in many places because the hyphen was a non-ASCII unicode character (perhaps an em dash), which caused a complaint under Python 2 because we don't set the encoding of our Python source files.

File: dark/diamond/conversion.py [hunk 1]
```diff
@@ -104,16 +105,27 @@ def records(self):
...
 
-    def saveAsJSON(self, fp):
+    def saveAsJSON(self, fp, writeBytes=False):
         """
         Write the records out as JSON. The first JSON object saved contains
         information about the DIAMOND algorithm.
 
         @param fp: A C{str} file pointer to write to.
+        @param writeBytes: If C{True}, the JSON will be written out as bytes
+            (not strings). This is required when fp is a bz2file.BZ2File.
```
Comment (bamueh) [inline dark/diamond/conversion.py]: `C{bz2file.BZ2File}`?

**Comment by bamueh** [comment]
Looks good. Just the one above.

File: dark/diamond/conversion.py (hunk 1 above)
Comment (terrycojones) [inline dark/diamond/conversion.py]: Fixed, thanks.

**Pull request merged**